### PR TITLE
feat(batcher): send response to clients once task has been confirmed in ethereum

### LINKED ISSUE #293: batcher: send response to clients once task has been confirmed in ethereum

### PR DESCRIPTION
**Pull request by entropidelic**
~~**NOTE**: #337 should be merged first~~

Up to now, the batcher sent batch inclusion response to clients as soon as the batch verification task was created in Ethereum. Now, the responses are delayed until the `BatchVerified` event is emitted, meaning that the task was verified by operators and signatures validated in Ethereum. A timeout for the awaiting was also added so that if it is exceeded, the connection with the pending clients is closed gracefully.

To test it:
1. start anvil with block time, aggregator, operator and batcher
2. for simplicity, install the batcher client with `make install_batcher_client`
3. Without moving from the root of the repo, run
```
batcher-client --proving_system GnarkPlonkBls12_381 \
--proof batcher/client/test_files/plonk_bls12_381/plonk.proof \
--public_input batcher/client/test_files/plonk_bls12_381/plonk_pub_input.pub \
--vk batcher/client/test_files/plonk_bls12_381/plonk.vk \
--repetitions 2
```
You should receive a response once the batch signatures are verified in etherem.
4. To test the timeout, now stop the aggregator and operator
5. Send some more proofs with
```
batcher-client --proving_system GnarkPlonkBls12_381 \
--proof batcher/client/test_files/plonk_bls12_381/plonk.proof \
--public_input batcher/client/test_files/plonk_bls12_381/plonk_pub_input.pub \
--vk batcher/client/test_files/plonk_bls12_381/plonk.vk \
--repetitions 4
```
Note that we are sending 4 so that we don't get the error about batch already submitted
6. After a minute, an error will be promted in the client terminal with information about the timeout and the connection will be closed

### CODE REVIEW

File: batcher/src/lib.rs [hunk 1]
```diff
@@ -318,14 +356,46 @@ async fn send_responses(
...
+    for (_, _, ws_sink) in finalized_batch.iter() {
+        let send_result = ws_sink.write().await.send(timeout_msg.clone()).await;
+        match send_result {
+            // When two or more proofs from the same client are included into a batch,
+            // there will be more than one `ws_sink` corresponding to that client. When one is
+            // closed, the other ones will raise this error. We can just ignore it.
+            Err(Error::Protocol(ProtocolError::SendAfterClosing)) => (),
+            Err(e) => {
+                error!("Error sending timeout response to clients: {}", e);
+                panic!();
```
Comment (taturosati) [inline batcher/src/lib.rs]: Dont panic, return error

File: batcher/src/lib.rs [hunk 2]
```diff
@@ -307,6 +326,25 @@ impl Batcher {
...
+        if let Ok(event) = event_result {
+            if &event.batch_merkle_root == batch_merkle_root {
+                info!("Batch operator signatures verified on Ethereum. Sending response to clients...");
+                break;
+            }
+        } else {
+            error!(
+                "Error awaiting for batch signature verification event: {}",
+                event_result.unwrap_err()
+            );
```
Comment (taturosati) [inline batcher/src/lib.rs]: Should probably return Result

**Review by jotabulacios — APPROVED** [review]
lgtm

**Review by taturosati — APPROVED** [review]
lgtm

**Review by MauroToscano — APPROVED** [review]

**Pull request merged**